From e5d03f9f3923e36e431403d8d82f8ebb1102ea23 Mon Sep 17 00:00:00 2001
From: John Zhuge <jzhuge@cloudera.com>
Date: Sat, 12 Aug 2017 00:43:19 -0700
Subject: [PATCH 2581/2748] HADOOP-14542. Add IOUtils.cleanupWithLogger that
 accepts slf4j logger API. Contributed by Chen
 Liang.

(cherry picked from commit b64951905e64f6fed581c28634be6ed15c190633)

Conflicts:
	hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/IOUtils.java

Change-Id: I99c053fa5909b03566c2d662bfc4346ab3473d1a
---
 .../main/java/org/apache/hadoop/io/IOUtils.java    |   24 +++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/IOUtils.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/IOUtils.java
index d4d5862..d8e8dd9 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/IOUtils.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/IOUtils.java
@@ -35,7 +35,7 @@
 import org.apache.hadoop.classification.InterfaceAudience;
 import org.apache.hadoop.classification.InterfaceStability;
 import org.apache.hadoop.conf.Configuration;
-import org.apache.hadoop.util.ChunkedArrayList;
+import org.slf4j.Logger;
 
 /**
  * An utility class for I/O related functionality. 
@@ -252,6 +252,28 @@ public static void cleanup(Log log, java.io.Closeable... closeables) {
   }
 
   /**
+   * Close the Closeable objects and <b>ignore</b> any {@link Throwable} or
+   * null pointers. Must only be used for cleanup in exception handlers.
+   *
+   * @param logger the log to record problems to at debug level. Can be null.
+   * @param closeables the objects to close
+   */
+  public static void cleanupWithLogger(Logger logger,
+      java.io.Closeable... closeables) {
+    for (java.io.Closeable c : closeables) {
+      if (c != null) {
+        try {
+          c.close();
+        } catch (Throwable e) {
+          if (logger != null) {
+            logger.debug("Exception in closing {}", c, e);
+          }
+        }
+      }
+    }
+  }
+
+  /**
    * Closes the stream ignoring {@link IOException}.
    * Must only be called in cleaning up from exception handlers.
    *
-- 
1.7.9.5

